### PR TITLE
fix: Stop redacting inside of reffed objects

### PR DESCRIPTION
**Pull request by tssweeney**
We had a production issue caused by https://github.com/wandb/weave/commit/0a78fcd7132b3e3387b41fd3f855695a1c7fd6db made it so that the eval calls lost their refs in inputs, which broke some un-tested contracts relied on by the UI. This fixes that code in a general way and adds specific tests for these issues.

I also verified in the UI that this generates the correct data and renders well in the unpatched UI

### CODE REVIEW

File: weave/weave_client.py [hunk 1]
```diff
@@ -818,6 +818,19 @@ def op_fn(*args, **kwargs):  # type: ignore
 
 
 def redact_sensitive_keys(obj: typing.Any) -> typing.Any:
+    # We should NEVER mutate reffed objects.
+    #
+    # 1. This code builds new objects that no longer have refs
+    # 2. Even if we did an in-place edit, that would invalidate the ref (since
+    # the ref is to the object's digest)
+
+    try:
```
Comment (jamie-rasmussen) [inline weave/weave_client.py:827]: Is there a reason we need this broad try/except?

File: weave/tests/test_evaluations.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    calls = client.server.calls_query(
+        tsi.CallsQueryReq(
+            project_id=client._project_id(),
+        )
+    )
+
+    # Flatten calls is a technique we use in the integration tests to assert the correct
+    # ordering of calls. This is used to assert that the calls are in the correct order
+    # as well as nested in the correct way. The returned list is the ordered list if calls
```
Comment (jamie-rasmussen) [inline weave/tests/test_evaluations.py:77]: if => of

**Review by jamie-rasmussen — APPROVED** [review]

**Review by nickpenaranda — APPROVED** [review]

File: weave/tests/test_evaluations.py (hunk 2 above)
Comment (tssweeney) [reply weave/tests/test_evaluations.py:77]: fixed

File: weave/weave_client.py (hunk 1 above)
Comment (tssweeney) [reply weave/weave_client.py:827]: I was under the impression this was risky, but you are correct, it should be safe. We actually recursively do this check immediately after this call so we should be good. Remove the try/except

File: weave/weave_client.py [hunk 3]
```diff
@@ -818,6 +818,15 @@ def op_fn(*args, **kwargs):  # type: ignore
 
 
 def redact_sensitive_keys(obj: typing.Any) -> typing.Any:
+    # We should NEVER mutate reffed objects.
+    #
+    # 1. This code builds new objects that no longer have refs
+    # 2. Even if we did an in-place edit, that would invalidate the ref (since
+    # the ref is to the object's digest)
+    ref = get_ref(obj)
+    if ref:
```
Comment (andrewtruong) [inline weave/weave_client.py:827]: nit:
```suggestion
    if get_ref(obj):
```

File: weave/tests/test_evaluations.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    calls = client.server.calls_query(
+        tsi.CallsQueryReq(
+            project_id=client._project_id(),
+        )
+    )
+
+    # Flatten calls is a technique we use in the integration tests to assert the correct
+    # ordering of calls. This is used to assert that the calls are in the correct order
+    # as well as nested in the correct way. The returned list is the ordered list of calls
+    # with the depth of each call.
```
Comment (andrewtruong) [inline weave/tests/test_evaluations.py:78]: this should be a docstring of the `flatten_calls` func

File: weave/tests/test_evaluations.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    res = []
+    for child in children:
+        res.append((child, depth))
+        res.extend(flatten_calls(calls, child.id, depth + 1))
+
+    return res
+
+
+def op_name_from_ref(ref: str) -> str:
+    return ref.split("/")[-1].split(":")[0]
```
Comment (andrewtruong) [inline weave/tests/test_evaluations.py:28]: should be in `refs.py`

File: weave/tests/test_evaluations.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        ("match_score2", 2),
+        ("Evaluation.predict_and_score", 1),
+        ("MyModel.predict", 2),
+        ("match_score1", 2),
+        ("match_score2", 2),
+        ("Evaluation.summarize", 1),
+    ]
+    assert got == exp
+
+    def isObjectRefWithName(val: str, name: str):
```
Comment (andrewtruong) [inline weave/tests/test_evaluations.py:100]:
```suggestion
def is_object_ref_with_name(val: str, name: str):
```

Comment (andrewtruong) [reply weave/tests/test_evaluations.py:100]: aside: are these utils generally useful?  Should we make them available as helpers inside `refs.py`?

**Review by andrewtruong — APPROVED** [review]

File: weave/weave_client.py (hunk 3 above)
Comment (tssweeney) [reply weave/weave_client.py:827]: fixed

File: weave/tests/test_evaluations.py (hunk 4 above)
Comment (tssweeney) [reply weave/tests/test_evaluations.py:78]: fixed

File: weave/tests/test_evaluations.py (hunk 6 above)
Comment (tssweeney) [reply weave/tests/test_evaluations.py:100]: fixed. Mostly a set of test utils that are probably common. can pull these out later

**Pull request merged**